File: src/so/aporia/module/settings/BooleanSetting.java
/*
 * Copyright (c) 2025-2026 BEVoid Project
 * Distributed under the BEVoid Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.module.settings;

import java.util.function.Supplier;

/**
 * Boolean (checkbox) setting.
 */
public class BooleanSetting extends Setting<Boolean> {
    
    public BooleanSetting(String name, String description, boolean defaultValue) {
        this(name, description, defaultValue, null);
    }
    
    public BooleanSetting(String name, String description, boolean defaultValue, Supplier<Boolean> visible) {
        super(name, description, defaultValue, visible);
    }
    
    public void toggle() { value = !value; }
    public boolean isEnabled() { return value; }
}

File: src/so/aporia/module/settings/BooleanSetting.kt
package so.aporia.module.settings
import so.aporia.utils.user.render.animation.SpringSimulator
import so.aporia.utils.user.render.core.AporiaRenderer
import so.aporia.module.impl.render.clickgui.ThemeManagerModule.Theme
import java.util.function.Supplier
import com.chaos.annotation.ChaosNative
@ChaosNative
class BooleanSetting @JvmOverloads constructor(
    name: String,
    description: String,
    defaultValue: Boolean,
    visible: Supplier<Boolean>? = null
) : Setting<Boolean>(name, description, defaultValue, visible) {

    private val toggleSpring = SpringSimulator(340f, 24f, if (defaultValue) 1f else 0f)
    private var lastTick = System.currentTimeMillis()

    val isEnabled: Boolean get() = value

    fun toggle() {
        value = !value
        toggleSpring.setTarget(if (value) 1f else 0f)
    }

    fun set(enabled: Boolean) {
        value = enabled
        toggleSpring.setTarget(if (value) 1f else 0f)
    }

    override fun draw(r: AporiaRenderer, x: Float, y: Float, w: Float, theme: Theme, mouseX: Float, mouseY: Float, isDropOpen: Boolean) {
        val now = System.currentTimeMillis()
        toggleSpring.update((now - lastTick) / 1000f)
        lastTick = now

        r.drawText("regular", name, x + 8f, y + (14f - 9f) / 2f - 1f, 9f, theme.guiSettingText)
        val tx = x + w - 26f
        val ty = y + (14f - 10f) / 2f
        val t = toggleSpring.value()
        r.drawRect(tx, ty, 18f, 10f, 5f,
            if (isEnabled) theme.guiEnabledDot else 0x66FFFFFF.toInt())
        val knobOff = t * 8f
        r.drawCircle(tx + knobOff + 5f, ty + 5f, 5f, -0x1)
        r.drawCircle(tx + knobOff + 5f, ty + 5f, 3f,
            if (isEnabled) 0xFF66DD66.toInt() else 0xFF888888.toInt())
    }
}